connection.go

```diff
@@ -1094,10 +1149,15 @@ func (conn *Connection) send(req Request, streamId uint64) *Future {
 		default:
 		}
 	}
+
+	atomic.AddInt64(&(conn.requestCnt), int64(1))
```

Comment:
1.1. newFuture checks conn.state != connShutdown.
2.2. processShutdown() sets up conn.state = connShutdown.
2.2. processShutdown() checks that atomic.LoadInt64(&(conn.requestCnt)) == 0 and starts reconnect().
1.2. The line increments conn.requestCnt.

As a result - we have a reconnect with in-progress request.

Reply:
Well, I can put this one line to be triggered a bit sooner, but I don't think it is possible to completely solve this one. Your request can always be too late.

Reply:
>  I don't think it is possible to completely solve this one

It will be enough to increase the counter before newFuture().

Reply:
Reconnect with in-progress request is possible without graceful shutdown. And, as far as I understand, behavior here should be the same -- either we spot on a closed connection (which is highly likely) and return a future with error in it, or we'll get ourselves a new connection to a new server (which is unlikely, especially in graceful shutdown. but possible) and execute it. Either way, there shouldn't be any deadlocks. The only question here is what is an "in-progress request". For now it's a request that has its future object returned at `Do()`'s end. Moreover, we guarantee that if a future has been created, it's counted by `requestCnt`(so there cannot be a future without its increment). The only thing we don't guarantee with current approach is that a request that **started** its `Do` is a "ready request". And it should be fine since we'll never be able to do this, even if it would be `Do`'s first line due to the asynchronous nature of the problem.

So everything should be ok there, and if by some accident it's not, then this bug has been existed for a long time since reconnect with in-progress request is a scenario that has been possible since this connector initial commits.

Reply:
> It will be enough to increase the counter before newFuture() to resolve the case.

Before the future has been put in the shard the request is not processed by anyone, so there should be no problems with incrementing its here. And based on the current code, this kind of treatment is already applied: here
https://github.com/tarantool/go-tarantool/blob/aac42004b51008a558deca8df7c6994d2769ce46/connection.go#L1140-L1151
we do not trigger `markDone()`, so if we want to increase the counter before `newFuture()`, we also need to manually decrease it there, which make code a bit more complicated while giving an unlucky request additional couple of microseconds to be considered "in-progress".

Reply:
In this case, the request will be submitted for send at the time of `shutdown` in progress due to the race condition. So we could send the request, but don't wait for a response.

This is not an old bug, but a new one (reconnect do not need to wait for completion of a request).

Reply:
1.1. newFuture checks conn.state != connShutdown.
2.1. shutdown() sets up conn.state = connShutdown.
2.2. shutdown() checks that atomic.LoadInt64(&(conn.requestCnt)) == 0 and starts reconnect().
1.2. The line increments conn.requestCnt
1.3. writer() goroutine sends the request.
2.3. shutdown() terminates the network connection.

Reply:
> This is not an old bug, but a new one

I don't get what you're talking about: there is no blocks disallowing to create a future while we're shutdowning in code before the counter:

https://github.com/tarantool/go-tarantool/blob/a1f28a6cd204d9ab6c3ea99f3192235cf3856f74/connection.go#L1305-L1322

Reply:
> submitted for send at the time of `shutdown`

`closeConnection` blocks all shards

Reply:
> newFuture checks conn.state != connShutdown.

Oh, really, I had missed that one. Thank you for pointing out, now I'll get it.

Reply:
But there is still no protection against reconnect started with another future while the first one is in any other line (except for shard blocks)

Reply:
> > submitted for send at the time of `shutdown`
> 
> `closeConnection` blocks all shards

1.1. newFuture checks conn.state != connShutdown.
2.1. shutdown() sets up conn.state = connShutdown.
2.2. shutdown() checks that atomic.LoadInt64(&(conn.requestCnt)) == 0 and starts reconnect().
1.2. The line increments conn.requestCnt
1.3. writer() goroutine sends the request.
2.3. shutdown()->reconnect()->closeConnection() blocks shards and terminates the network connection.

Reply:
Moved the increment